packages/contracts-bedrock/src/L1/OPContractsManager.sol

```diff
@@ -562,37 +542,136 @@ contract OPContractsManagerUpgrader is OPContractsManagerBase {
-            IFaultDisputeGame permissionlessDisputeGame = IFaultDisputeGame(
-                address(getGameImplementation(IDisputeGameFactory(opChainAddrs.disputeGameFactory), GameTypes.CANNON))
+            // Start by upgrading the SystemConfig contract to have the l2ChainId.
+            upgradeToAndCall(
+                _opChainConfigs[i].proxyAdmin,
+                address(_opChainConfigs[i].systemConfigProxy),
+                impls.systemConfigImpl,
+                abi.encodeCall(ISystemConfig.upgrade, (l2ChainId))
             );
 
-            if (address(permissionlessDisputeGame) != address(0)) {
-                // Deploy and set a new permissionless game to update its prestate
+            // Separate context to avoid stack too deep.
+            IAnchorStateRegistry newAnchorStateRegistryProxy;
+            {
+                // Deploy a new AnchorStateRegistry contract.
+                // We use the SOT suffix to avoid CREATE2 conflicts with the existing ASR.
+                newAnchorStateRegistryProxy = IAnchorStateRegistry(
+                    deployProxy({
+                        _l2ChainId: l2ChainId,
+                        _proxyAdmin: _opChainConfigs[i].proxyAdmin,
+                        _saltMixer: reusableSaltMixer(_opChainConfigs[i]),
+                        _contractName: "AnchorStateRegistry-SOT"
+                    })
+                );
+
+                // Separate context to avoid stack too deep.
+                {
+                    // Get the existing anchor root from the old AnchorStateRegistry contract.
+                    // Get the AnchorStateRegistry from the PermissionedDisputeGame.
+                    (Hash root, uint256 l2BlockNumber) = getAnchorStateRegistry(
+                        IFaultDisputeGame(address(permissionedDisputeGame))
+                    ).anchors(respectedGameType);
+
+                    // Upgrade and initialize the AnchorStateRegistry contract.
+                    // Since this is a net-new contract, we need to initialize it.
+                    upgradeToAndCall(
+                        _opChainConfigs[i].proxyAdmin,
+                        address(newAnchorStateRegistryProxy),
+                        impls.anchorStateRegistryImpl,
+                        abi.encodeCall(
+                            IAnchorStateRegistry.initialize,
+                            (
+                                superchainConfig,
+                                IDisputeGameFactory(opChainAddrs.disputeGameFactory),
+                                OutputRoot({ root: root, l2BlockNumber: l2BlockNumber }),
+                                respectedGameType
+                            )
+                        )
+                    );
+                }
+            }
+
+            // Upgrade the OptimismPortal contract implementation.
+            upgradeTo(_opChainConfigs[i].proxyAdmin, opChainAddrs.optimismPortal, impls.optimismPortalImpl);
```

Comment:
Nit, this is fine, but you could save gas/loc/bytecode by doing this as an `upgradeToAndCall()` at the same time as the `portal.upgrade()` call below.

Reply:
Gotcha, since it's technically fine will leave as-is to reduce further changes to this PR